front/lib/api/assistant/pubsub.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  AgentActionEvent,
+  AgentActionSuccessEvent,
+  AgentErrorEvent,
+  AgentGenerationSuccessEvent,
+} from "@app/lib/api/assistant/agent";
+import {
+  AgentMessageNewEvent,
+  postUserMessage,
+  UserMessageNewEvent,
+} from "@app/lib/api/assistant/conversation";
+import { GenerationTokensEvent } from "@app/lib/api/assistant/generation";
+import { Authenticator } from "@app/lib/auth";
+import { redisClient } from "@app/lib/redis";
+import logger from "@app/logger/logger";
+import {
+  ConversationType,
+  Mention,
+  UserMessageContext,
+} from "@app/types/assistant/conversation";
+
+export async function postUserMessageWithPubSub(
+  auth: Authenticator,
+  {
+    conversation,
+    message,
+    mentions,
+    context,
+  }: {
+    conversation: ConversationType;
+    message: string;
+    mentions: Mention[];
+    context: UserMessageContext;
+  }
+) {
+  const redis = await redisClient();
+  try {
+    for await (const event of postUserMessage(auth, {
+      conversation,
+      message,
+      mentions,
+      context,
+    })) {
+      switch (event.type) {
+        case "user_message_new":
+        case "agent_message_new": {
+          const pubsubChannel = getConversationChannelId(conversation.sId);
+          await redis.xAdd(pubsubChannel, "*", {
+            payload: JSON.stringify(event),
+          });
+          break;
+        }
+        case "retrieval_params":
+        case "agent_error":
+        case "agent_action_success":
+        case "retrieval_documents":
+        case "generation_tokens":
+        case "agent_generation_success":
+        case "agent_message_success": {
+          const pubsubChannel = getMessageChannelId(event.messageId);
+          await redis.xAdd(pubsubChannel, "*", {
+            payload: JSON.stringify(event),
+          });
+          break;
+        }
+
+        default:
+          ((blockParent: never) => {
+            logger.error("Unknown event type", blockParent);
+          })(event);
+          return null;
+      }
+    }
+  } finally {
+    await redis.quit();
+  }
+  console.log("exiting postUserMessageWithPubSub", message, conversation.sId);
+}
+
+export async function* getConversationEvents(
+  conversationId: string,
+  lastEventId: string | null
+): AsyncGenerator<{
+  eventId: string;
+  data: UserMessageNewEvent | AgentMessageNewEvent;
+}> {
+  const redis = await redisClient();
+  const pubsubChannel = getConversationChannelId(conversationId);
+
+  try {
+    while (true) {
+      const events = await redis.xRead(
+        { key: pubsubChannel, id: lastEventId ? lastEventId : "0-0" },
+        // weird, xread does not return on new message when count is = 1. Anything over 1 works.
```

Comment:
Hey! Reading the code to prepare, thanks for all that aric :pray: 

NRE - Just asking : something I don't understand here: if we use count = 1 then only 1 event on the array max right, and no need for the loop below? Why not get e.g. count = 100, to limit the calls to redis -- and keep the loop?

Reply:
Sorry to drop in uninvited :sweat_smile: hope I'm not disturbing

Reply:
For context: this comment was wrong, count is honored, I am not sure exactly if redis can return more than 1 event when using the blocking call because I have read that it returns on the first one, but for the first call with no "lastEventId" set, it does return multiple rows.

In any case, that's the redis interface, they always return an array, so we'll always have to loop. We could fetch event[0] but it does not change the code much I think. Does it?